src/demProfiler/index.ts

```diff
@@ -0,0 +1,237 @@
+import type { Feature, LineString } from "geojson";
+
+export * from "./getElevation";
+export * from "./tileCover";
+
+export interface TileImage {
+  channels: 1 | 2 | 3 | 4;
+  image: Uint8ClampedArray;
+}
+
+export type TileRequest = (
+  x: number,
+  y: number,
+  zoom: number
+) => Promise<TileImage>;
+
+export interface Options {
+  /** type of metric to use. Meters or feet. Default: meters */
+  metric?: "m" | "ft";
+  /** Zoom that is queried from the server. Default: 13 */
+  zoom?: number;
+  /** Tile size of the images returned. Default 512 */
+  tileSize?: number;
+  /** Tile Request method */
+  tileRequest: TileRequest;
+  /** Elevation parser. Default: elevation = -10000 + ((R * 256 * 256 + G * 256 + B) * 0.1) */
+  elevationParser?: ElevationParser;
+  /** smooth out the elevation to make the visual aesthetic nicer */
+  smooth?: boolean;
+}
+
+export interface ElevPoint {
+  /** Distance along path from the starting point in the metric defind by options */
+  distance: number;
+  /** Elevation of the point */
+  elevation: number;
+  /** Coordinates of the point in lat-lon */
+  coordinate: [lon: number, lat: number];
+  /** Tile that the point is in */
+  tile: TileID;
+}
+
+export interface Output {
+  /** Total length of the path in the metric defined by options */
+  distance: number;
+  /** Minimum elevation of the path */
+  minElevation: number;
+  /** Maximum elevation of the path */
+  maxElevation: number;
+  /** Average elevation of the path */
+  avgElevation: number;
+  /** Elevation at the start of the path */
+  startElevation: number;
+  /** Elevation at the end of the path */
+  endElevation: number;
+  /** Array of points along the path */
+  points: ElevPoint[];
+}
+
+/**
+ * Given a GeoJSON LineString or Feature<Linestring>, return the elevation data for the path.
+ * This algorithm will automatically break the path into denser segments relative to the zoom level if necessary.
+ */
+export async function profileLineString(
+  path: Feature<LineString> | LineString,
+  options: Options
+): Promise<Output> {
+  // get the tile cover
+  const coordinates =
+    "geometry" in path ? path.geometry.coordinates : path.coordinates;
+  const { coords, tiles } = tileCover(
+    coordinates,
+    options.zoom ?? 13,
+    options.tileSize ?? 512
+  );
+  // get tiles
+  const tileCache = await getTiles(tiles, options.tileRequest);
+  // get the elevation data
+  let points = getElevations(
+    coords,
+    tileCache,
+    options.tileSize ?? 512,
+    options.elevationParser
+  );
+  // smooth the elevation data if needed
+  if (options.smooth === true) points = smoothElevation(points);
+  // calculate the output
+  let output = buildOutput(points);
+  // convert to miles if needed
+  if (options.metric === "ft") output = toFeet(output);
+
+  return output;
+}
+
+/** Request all the tiles we need */
+async function getTiles(
+  tiles: TileID[],
+  tileRequest: TileRequest
```

Comment:
Since this is part of MapTiler client lib, we should probably give a default to fetching our actual elevation tileset, using the apiKey stored in the client config (config.apiKey). But for tests or custom settings, making it optional is still nice.

Reply:
see the other comment on this topic